### DISCUSSION TITLE
RLS not working for inter schema query

### DISCUSSION THREAD

**Discussion by softmarshmallow** [supabase/supabase#21758] [Questions]
# Bug report

- [ x ] I confirm this is a bug with Supabase, not with my own application. 
- [ x ] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

I have encountered an issue with Row-Level Security (RLS) policies not behaving as expected. Despite confirming the correctness of the policy's logic through manual testing, the RLS policy does not authorize access as it should when applied.

## To Reproduce

Given the following database schema setup:

- Schemas: `public`, `app`, and `auth` (for user authentication).
- Tables involved:
  - `public.project` (links to `organization`)
  - `public.organization` (linked from `project`, linked to `organization_member`)
  - `public.organization_member` (links users to `organization`)
  - `app.form` (contains `project_id` linking to `public.project`)

The RLS policy is intended to allow users to access rows from the `app.form` table if they are members of the corresponding project's organization. The policy logic is as follows:

```sql
CREATE POLICY allow_read_for_project_organization_member
ON app.form
FOR SELECT
USING (
    EXISTS (
        SELECT 1
        FROM public.project
        INNER JOIN public.organization ON public.organization.id = public.project.organization_id
        INNER JOIN public.organization_member ON public.organization_member.organization_id = public.organization.id
        WHERE public.project.id = app.form.project_id
          AND public.organization_member.user_id = auth.uid()
    )
);
```

1. Apply the above RLS policy to the app.form table.
2. Attempt to query the app.form table as a user who is a member of the relevant organization (confirmed through a manual query that returns true for the user being a member).
3. The query does not return the expected rows, despite the user meeting the RLS policy conditions.

Manual query for testing:

```
SELECT EXISTS (
  SELECT 1
  FROM public.project
  JOIN public.organization ON public.organization.id = public.project.organization_id
  JOIN public.organization_member ON public.organization_member.organization_id = public.organization.id
  WHERE public.project.id = <specific_project_id>
    AND public.organization_member.user_id = '<user_id>'
) AS user_is_member_of_organization;

```

This manual query confirms the user is indeed a member of the organization associated with a specific project, yet the RLS policy does not authorize access as it should.

## Expected behavior

I expected the RLS policy to authorize the user to access rows from the app.form table based on their membership in the relevant organization, as confirmed by the manual test query.

## Additional context

This issue has led to significant debugging efforts, and manual verification of the policy's logic suggests it should work as intended. I am seeking insights into why the RLS policy might not be functioning correctly in this scenario.

**Comment by GaryAustin1** [comment]
Moving to this to a discussion for now as not clear it is a bug (unlikely as Postgres is pretty solid) and you are seeking debug help.
What does not working mean?  An error?  No data returned?
Are you testing this from the API or in the SQL editor with user impersonation?
Do you meet select access to all the tables involved in the query?
RLS will not grant access to a schema.  Does the user role have access to the schema?  Do the API settings allow access to app schema?  Once again knowing the failure mode would help.

**Comment by softmarshmallow** [comment]
- No data returned without an error (other quickstart rls template works)
- Via supabase-js client, with authorized user (only one user in database for testing purpose, pretty solid)
- Yes, the custom schema is allowed on api settings (and said above, works with quickstart rls on the app schema)
- "Do you meet select access to all the tables involved in the query?" - no only for the target row on `app.form`


It's not a debug request, I assume It's a bug considering the query itself works (returns true) but once used inside a rls, it does not return any row

**Reply by GaryAustin1** [reply]
You have to meet select access for all tables you use in RLS if called from the API.
From the SQL editor you do not.  There is also an impersonation mode at bottom right of the SQL editor where you can test as an authenticated or anon user.
